File: HW/HW12/Hw12.java
import java.util.*;

class Graph2 {
    public int minCostToSupplyWater(int n, int[] wells, int[][] pipes) {
        List<int[]> allEdges = new ArrayList<>();
        for (int i = 0; i < wells.length; i++) allEdges.add(new int[]{0, i + 1, wells[i]});
        for (int[] pipe : pipes) allEdges.add(pipe);
        allEdges.sort(Comparator.comparingInt(a -> a[2]));

        int[] parent = new int[n + 1];
        for (int i = 0; i <= n; i++) parent[i] = i;

        int minCost = 0;
        for (int[] edge : allEdges) {
            int root1 = find(parent, edge[0]), root2 = find(parent, edge[1]);
            if (root1 != root2) { parent[root1] = root2; minCost += edge[2]; }
        }

        return minCost;
    }

    private int find(int[] parent, int x) {
        return parent[x] == x ? x : (parent[x] = find(parent, parent[x]));
    }

    public static void main(String[] args) {
        Graph2 graph = new Graph2();
        int result = graph.minCostToSupplyWater(2, new int[]{1, 1}, new int[][]{{1, 2, 1}, {1, 2, 2}});
        System.out.println(result); 
    }
}

File: HW/HW12/HW12.py
from typing import List

class Graph:
    def minCostToSupplyWater(self, n: int, wells: List[int], pipes: List[List[int]]) -> int:
        all_edges = []
        
        for i, cost in enumerate(wells):
            all_edges.append((0, i+1, cost)) 
        
        for pipe in pipes:
            all_edges.append((pipe[0], pipe[1], pipe[2]))
        
        all_edges.sort(key=lambda x: x[2])
        
        parent = [i for i in range(n+1)]
        
        def find(x):
            if parent[x] != x:
                parent[x] = find(parent[x])
            return parent[x]

        min_cost = 0
        for edge in all_edges:
            house1, house2, cost = edge
            root1 = find(house1)
            root2 = find(house2)
            if root1 != root2:  
                parent[root1] = root2
                min_cost += cost
        
        return min_cost

graph = Graph()
print(graph.minCostToSupplyWater(2, [1,1], [[1,2,1],[1,2,2]])) 
